src/wallet/KintoWallet.sol

```diff
@@ -351,34 +373,26 @@ contract KintoWallet is Initializable, BaseAccount, TokenCallbackHandler, IKinto
         dest.functionCallWithValue(func, value);
     }
 
-    // Function to extract the first target contract
-    function _getAppContract(bytes calldata callData) private view returns (address) {
-        // Extract the function selector from the callData
-        bytes4 selector = bytes4(callData[:4]);
+    // extracts `target` contract from the callData
+    // @dev when calling executeBatch, we use the first element of the targets to determine the app
+    function _getTarget(bytes calldata callData) private view returns (address target) {
+        bytes4 selector = bytes4(callData[:4]); // extract the function selector from the callData
 
-        // Compare the selector with the known function selectors
         if (selector == IKintoWallet.executeBatch.selector) {
-            // Decode callData for executeBatch
+            // decode executeBatch callData
             (address[] memory targets,,) = abi.decode(callData[4:], (address[], uint256[], bytes[]));
-            address lastTargetContract = appRegistry.getSponsor(targets[targets.length - 1]);
+            require(targets.length > 0, "KW: no targets");
+
+            // find the first non-wallet contract in the batch to determine the target
             for (uint256 i = 0; i < targets.length; i++) {
-                // App signer should only be valid for the app itself and its children
-                // It is important that wallet calls are not allowed through the app signer
-                if (!appRegistry.isSponsored(lastTargetContract, targets[i])) {
```

Comment:
are we safe to remove this from here? the paymaster is checking this as well

Reply:
we cannot remove it. The Sponsor paymaster should allow calls to the wallet but here we definitely cannot

Reply:
the logic is not the same in both places

Reply:
but i'm doing the check for that here: https://github.com/KintoXYZ/kinto-core/pull/47/files#diff-8fe9b6b2a7af2cbca0f7c947042eb63d079351c559c0868d38b491e28d8bb600R294